x/yarpctest/request_unary.go

```diff
@@ -113,13 +103,36 @@ func GRPCRequest(options ...api.RequestOption) api.Action {
 		require.NoError(t, out.Start())
 		defer func() { assert.NoError(t, out.Stop()) }()
 
+		sendRequestAndValidateResp(t, out, opts)
+	})
+}
+
+func sendRequestAndValidateResp(t testing.TB, out transport.UnaryOutbound, opts api.RequestOpts) {
+	f := func(i int) bool {
 		resp, cancel, err := sendRequest(out, opts.GiveRequest, opts.GiveTimeout)
 		defer cancel()
-		validateError(t, err, opts.WantError)
-		if opts.WantError == nil {
-			validateResponse(t, resp, opts.WantResponse)
+
+		if i == opts.AttemptCount-1 {
+			validateError(t, err, opts.WantError)
+			if opts.WantError == nil {
+				validateResponse(t, resp, opts.WantResponse)
+			}
+			return true
 		}
-	})
+
+		if err != nil || matchResponse(resp, opts.WantResponse) != nil {
```

Comment:
The idea is that, we validate the error and responses ONLY on the last retry.

It's possible that request returns a nil error, but response doesn't match. If we use what you suggest here, `validateResponse` would fail immediately without future retry, which is not what we want.

Reply:
ahh yes my mistake!